File: go/anyqmw/strategy.go
package anyqmw

import (
	"context"
	"errors"

	"github.com/sns45/anyq/go/core"
)

// StrategyName is the companion strategy's stable name. It is not one of anyq's park-free names, so a consumer
// on an adapter without native delay applies its park-downgrade policy to it (set AllowParkDowngrade there).
const StrategyName = "anyonce-idempotency"

// Strategy is REQ-Q-8 and Q2: anyq's dead-letter and delay primitives are consumer hooks reachable only through
// a decision, so Wrap returns typed errors and this strategy translates them. An in-flight duplicate parks for
// the lease remainder, a payload mismatch dead-letters with reason fingerprint-mismatch, and everything else
// goes to inner, which defaults to anyq's reference RetryThenDeadLetter when nil.
func Strategy(inner core.Strategy) core.Strategy {
	delegate := inner
	if delegate == nil {
		delegate = core.RetryThenDeadLetter(nil)
	}
	return core.Custom(StrategyName, func(ctx context.Context, sc core.StrategyContext) (core.Decision, error) {
		var inFlight *InFlightError
		if errors.As(sc.Err, &inFlight) {
			inFlight.MarkTranslated()
			return core.Park(inFlight.DelayMs), nil
		}
		if errors.Is(sc.Err, ErrFingerprintMismatch) {
			return core.DeadLetter("fingerprint-mismatch"), nil
		}
		return delegate.Decide(ctx, sc)
	})
}

File: packages/anyq/src/strategy.ts
import {
  type RetryDecision,
  type RetryStrategy,
  type RetryStrategyContext,
  retryThenDeadLetter,
} from '@anyq/core';
import { isFingerprintMismatchError, isInFlightError } from './errors';

/** The strategy's stable name. It is not one of anyq's park-free names, so a park-capable adapter check sees it. */
export const IDEMPOTENCY_STRATEGY_NAME = 'anyonce-idempotency';

/**
 * REQ-Q-8 and Q2: anyq's dead-letter and delay primitives are consumer hooks reachable only through a strategy
 * decision, so the door throws typed errors and this strategy translates them. An in-flight duplicate parks for
 * the lease remainder; a payload mismatch dead-letters with reason fingerprint-mismatch; everything else is the
 * inner strategy's business, which defaults to anyq's reference retryThenDeadLetter.
 *
 * On adapters without native delayed redelivery (Kafka, Redis Streams) anyq downgrades the park to an in-process
 * sleep followed by a re-invocation, which re-enters begin after the lease has expired. Those consumers must set
 * allowParkDowngrade so anyq's park policy check permits the downgrade.
 */
export function idempotencyStrategy<T = unknown>(inner?: RetryStrategy<T>): RetryStrategy<T> {
  const delegate = inner ?? retryThenDeadLetter<T>();
  return {
    name: IDEMPOTENCY_STRATEGY_NAME,
    async decide(ctx: RetryStrategyContext<T>): Promise<RetryDecision> {
      const error: unknown = ctx.error;
      if (isInFlightError(error)) {
        error.translated = true;
        return { action: 'park', delayMs: error.delayMs };
      }
      if (isFingerprintMismatchError(error)) {
        return { action: 'deadLetter', reason: 'fingerprint-mismatch' };
      }
      return delegate.decide(ctx);
    },
  };
}
